src/peripheral/dcb.rs

```diff
@@ -20,18 +20,32 @@ pub struct RegisterBlock {
 }
 
 impl DCB {
-    /// Enables TRACE. This is for example required by the 
+    /// Enables TRACE. This is for example required by the
     /// `peripheral::DWT` cycle counter to work properly.
     /// As by STM documentation, this flag is not reset on
     /// soft-reset, only on power reset.
     pub fn enable_trace(&mut self) {
         // set bit 24 / TRCENA
         unsafe { self.demcr.modify(|w| w | DCB_DEMCR_TRCENA); }
     }
-    
+
     /// Disables TRACE. See `DCB::enable_trace()` for more details
     pub fn disable_trace(&mut self) {
         // unset bit 24 / TRCENA
         unsafe { self.demcr.modify(|w| w & !DCB_DEMCR_TRCENA); }
     }
+
+    /// Is there a debugger attached? (see notes)
+    ///
+    /// Note 1: This function is [reported not to
+    /// work](http://web.archive.org/web/20180821191012/https://community.nxp.com/thread/424925#comment-782843)
+    /// on Cortex-M0 devices. Per the ARM v6-M Architecture Reference Manual, "Access to the DHCSR
+    /// from software running on the processor is IMPLEMENTATION DEFINED". Indeed, from the
+    /// [Cortex-M0+ r0p1 Technical Reference Manual](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0484c/BABJHEIG.html), "Note Software cannot access the debug registers."
+    ///
+    /// Note 2: This function reads the DHCSR register, and therefore clears S_RESET_ST and
```

Comment:
I'm still not happy about this being a static method that modifies the contents of the register. The documentation says:

> Writes to this register in any size other than word are Unpredictable. It is acceptable to read in any size, and you can use it to avoid or intentionally change a sticky bit.

Could you turn this into a byte read to avoid clearing the S_RESET_ST and S_RETIRE_ST bits? Then it'd be fine to keep this as a static method.

Reply:
I've tried a couple of ways to do this, but am not sure which you'd prefer.

Byte-wise access:
```rust
pub struct RegisterBlock {
    /// Debug Halting Control and Status
    pub dhcsr: [RW<u8>;4],
...
```

breaks writing to the register, which requires that the high halfword always be `0xA05F`. 

Using an untagged union:
```rust
pub union DHCSR {
    /// 32-bit word access
    w: WO<u32>,
    /// Individual byte access
    b: [RW<u8>; 4],
}

/// Register block
#[repr(C)]
pub struct RegisterBlock {
    /// Debug Halting Control and Status
    pub dhcsr: DHCSR,
...
```
looks like a good solution, but untagged unions of non-Copy types is unstable.

I could try to do some unsafe trickery by doing pointer math but that seems like a bad idea. What do you think?

As an aside, I'm sorry this tiny little change has eaten up so much energy. It seemed so simple at first. :(

Reply:
I was reminded of this today - any suggestions? I'm happy to go with whatever's best, I just don't know what it is. :)

Reply:
@thenewwazoo sorry for dropping the ball here

> breaks writing to the register, which requires that the high halfword always be 0xA05F.

this would be a breaking change; also it breaks writes as you say

> Using an untagged union:

this is unstable so it's a no-go

> I could try to do some unsafe trickery by doing pointer math but that seems like a bad idea. 

using pointers seems like the only stable, backward-compatible solution. You can write something like `(*(&dhcsr as *const [RW<u8>; 4]))[index].read())` to avoid the pointer math. Though pointer math is probably easier to read `ptr::read_volatile((&dhcsr as *const u8).add(index))`.